### ISSUE TITLE
net/http: server not setting Content-Length in certain cases

### ISSUE THREAD

**Issue by muirdm** [golang/go#44953]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.16.2 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes.

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/muir/.cache/go-build"
GOENV="/Users/muir/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/Users/muir/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/Users/muir/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.16.2"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="[redacted]"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1046132156=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

https://play.golang.org/p/w-t_Fq5N2JD (not runnable ATM due to playground issue).

### What did you expect to see?

"Content-Length" set to 1 in the response, as in Go 1.15.

### What did you see instead?

"Content-Length" not set in the response.

**Comment by toothrot** [comment]
/cc @neild 

I'll look into the playground timeouts separately.

**Comment by muirdm** [comment]
This seems to be incorrect behavior in Go 1.16 based on the documented http.ResponseWriter.Write behavior: `if the total size of all written data is under a few KB and there are no Flush calls, the Content-Length header is added automatically.`.

```
	// If WriteHeader has not yet been called, Write calls
	// WriteHeader(http.StatusOK) before writing the data. If the Header
	// does not contain a Content-Type line, Write adds a Content-Type set
	// to the result of passing the initial 512 bytes of written data to
	// DetectContentType. Additionally, if the total size of all written
	// data is under a few KB and there are no Flush calls, the
	// Content-Length header is added automatically.
```

**Comment by neild** [comment]
This is caused by [CL 249238](https://golang.org/cl/249238), which changes `ResponseWriter.ReadFrom` to issue a single 512 byte read for `Content-Type` detection, where previously it used `io.Copy` with a 512-byte `io.LimitReader`.

I think this is a bug in that CL; content-detection should read up to 512 bytes, issuing multiple reads if necessary.

**Comment by networkimprov** [comment]
@gopherbot please backport to 1.16; this is a regression.

**Comment by mknyszek** [comment]
Any updates on landing the fix to this? It's still in the Go 1.17 milestone, and this appears to be a regression. Its backport has already been approved.

**Comment by networkimprov** [comment]
cc @fraenkel @bradfitz
